// @bravocms/autosaving/autosaving.js
/*
 * jQuery plugin that makes inputs save own content on the fly
 */
(function (global, factory) {
    if (typeof exports === 'object' && typeof module !== 'undefined') {
        require('timer');
        factory(exports, require('jquery'));
    } else if (typeof define === 'function' && define.amd) {
        define(['exports', 'jquery', 'timer'], factory);
    } else {
        factory((global.autosaving = {}), global.jQuery);
    }
}(this, (function (exports, $) {
    var autosaving = $(function ($) {

        $(document.body).on('autosaving', '[autosaving-name]', function (e) {
            console.log('saving...', $(this).attr('name'));

            var data = {};
            var value;
            if ($(this).attr('autosaving-sync')) {
                $("[autosaving-sync='" + $(this).attr('autosaving-sync') + "']").each(function () {
                    value = $(this).val();
                    var name = $(this).attr('autosaving-name') ? $(this).attr('autosaving-name') : $(this).attr('name');

                    if ($(this).attr('type') === 'checkbox' && !$(this).prop('checked')) {
                        if (typeof data[name] !== 'undefined') {
                            return;
                        }

                        value = null;
                    }

                    data[name] = value;
                });
            } else {
                value = $(this).val();

                if ($(this).attr('type') === 'checkbox' && !$(this).prop('checked')) {
                    value = null;
                }

                if ($(this).attr('autosaving-name')) {
                    data[$(this).attr('autosaving-name')] = value;
                } else if ($(this).attr('name')) {
                    data[$(this).attr('name')] = value;
                }
            }

            if ($(this).data('actualValue') === value) {
                // skip saving
                $(this).removeClass('autosaving').removeClass('autosaving-done');
                if ($(this).attr('autosaving-handle')) {
                    $($(this).attr('autosaving-handle')).removeClass('autosaving').removeClass('autosaving-done');
                }

                $(this).stopTime('autosaving');
                $(this).stopTime('autosaving-done');

                console.log('skip on actualValue:', data);
                return;
            }

            // mark element as saving
            $(this).addClass('autosaving').removeClass('autosaving-done');
            if ($(this).attr('autosaving-handle')) {
                $($(this).attr('autosaving-handle')).addClass('autosaving').removeClass('autosaving-done');
            }

            // Stop previous saving, Start saving with delay
            $(this).stopTime('autosaving').oneTime(1500, 'autosaving', function () {

                //console.log('saving');
                var url;
                var method = 'POST';

                if ($(this).attr('autosaving-url')) {
                    url = $(this).attr('autosaving-url');
                } else {
                    url = $(this).closest('form').attr('action');
                }

                if ($(this).attr('autosaving-method')) {
                    method = $(this).attr('autosaving-method');
                } else if ($(this).closest('form').attr('method')) {
                    method = $(this).closest('form').attr('method');
                }

                var post = $(this).data('post');
                if (post) {
                    var vars = post.split('&');
                    for (var i = 0; i < vars.length; i++) {
                        var splitOn = vars[i].indexOf('=');
                        if (splitOn > 0) {
                            var postKey = vars[i].slice(0, splitOn);
                            var postValue = vars[i].slice(splitOn + 1);

                            data[decodeURIComponent(postKey)] = decodeURIComponent(postValue);
                        }
                    }
                }

                console.log('saving:', data);
                // ajax save
                $.ajax({
                    url: url,
                    method: method,
                    data: data,
                    context: this,
                    dataType: 'jsonp',
                    success: function (data) {
                        console.log('saved', data);
                        if (data.redirect) {
                            window.location = data.redirect;
                        }

                        if (data.refresh) {
                            window.location = window.location;
                        }

                        if ($(this).attr('autosaving-handle')) {
                            $($(this).attr('autosaving-handle')).removeClass('autosaving').addClass('autosaving-done');
                        }
                        $(this).removeClass('autosaving').addClass('autosaving-done').trigger('saved.autosaving', [data]);
                        $(this).data('actualValue', value);

                        $(this).stopTime('autosaving-done').oneTime(1500, 'autosaving-done', function () {
                            if ($(this).attr('autosaving-handle')) {
                                $($(this).attr('autosaving-handle')).removeClass('autosaving-done');
                            }
                            $(this).removeClass('autosaving-done');
                        });
                    },
                    error: function (data) {
                        console.log('error', data);
                        if ($(this).attr('autosaving-handle')) {
                            $($(this).attr('autosaving-handle')).removeClass('autosaving');
                        }
                        $(this).removeClass('autosaving');
                    }
                });
            });
        });

        // trigger autosaving on keyup
        $(document.body).on('keyup', '[autosaving-name]', function (e) {
            $(this).trigger('autosaving');
        });

        // trigger autosaving on change
        $(document.body).on('change', '[autosaving-name]', function (e) {
            $(this).trigger('autosaving');
        });

    });
})));
